Source/WebCore/page/Quirks.cpp

```diff
@@ -1635,4 +1635,35 @@ bool Quirks::needsConfigurableIndexedPropertiesQuirk() const
     return needsQuirks() && m_needsConfigurableIndexedPropertiesQuirk;
 }
 
+// Canvas fingerprinting (rdar://107564162)
+bool Quirks::shouldEnableCanvas2DAdvancedPrivacyProtectionQuirk() const
+{
+    if (!needsQuirks() || !m_document->noiseInjectionHashSalt())
+        return false;
+    auto& url = m_document->topDocument().url();
+    auto host = url.host();
+    auto path = url.path();
+
+    if (!equalLettersIgnoringASCIICase(host, "walgreens.com"_s) && !host.endsWithIgnoringASCIICase(".walgreens.com"_s)
+        && !equalLettersIgnoringASCIICase(host, "fedex.com"_s) && !host.endsWithIgnoringASCIICase(".fedex.com"_s))
+        return false;
+
+    if ((host.endsWithIgnoringASCIICase("walgreens.com"_s) && equalLettersIgnoringASCIICase(path, "/login.jsp"_s))
+        || (host.endsWithIgnoringASCIICase("fedex.com"_s) && path.startsWithIgnoringASCIICase("/secure-login/"_s))
+        || (host.endsWithIgnoringASCIICase("fedex.com"_s) && path.startsWithIgnoringASCIICase("/fedextrack/"_s)))
```

Comment:
Nit - would be better to avoid the redundant `endsWith` check by adjusting the if statement:

```suggestion
        || (host.endsWithIgnoringASCIICase("fedex.com"_s) && (path.startsWithIgnoringASCIICase("/secure-login/"_s) || path.startsWithIgnoringASCIICase("/fedextrack/"_s)))
```

(or by pulling it out into a separate local variable)